drivers: analog-io: max2201x : Add MAX2201X driver support

RaduSabau1 opened the pull request:
## Pull Request Description

The MAX22017 is a two-channel industrial-grade 
configurable analog output device that can be changed 
on-the-fly in software to voltage or current output mode. 
The MAX22017 offers one 16-bit DAC per output 
channel. The MAX22017 can operate from an internal or 
external reference.

The first commit consists of main driver header and source files.
The second commit consists of IIO driver header and source files.
The third commit consists of driver's README.rst documentation file.
The fourth commit consists of project related files.
The fifth commit consists of project's README.rst documentation file.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

RaduSabau1 commented:
V2

- Fixed CI errors.

amiclaus reviewed (commented):
there are still some doc build issues that need handling.

RaduSabau1 commented:
V3

- Fixed doc build issues.

rbolboac commented on drivers/analog-io/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,919 @@
+static int max2201x_iio_reg_write(struct max2201x_iio_desc *iio_desc,
+				  uint32_t reg, uint32_t writeval);
+
+static struct iio_attribute max2201x_attrs[] = {
```
regarding attributes, the names should be compatible with attributes name from Linux IIO sub-system, for compatibility reasons, so please take a look at this list: https://github.com/torvalds/linux/blob/master/drivers/iio/industrialio-core.c#L155 and only leave out the attributes available there. the other attributes should be moved in max2201x_debug_attrs

rbolboac commented on drivers/analog-io/max2201x/README.rst:
```diff
@@ -115,7 +115,7 @@ Device Attributes
 
 Channel Attributes
------------------
+------------------
```
does not look related to this commit

rbolboac commented on projects/max2201x/builds.json:
```diff
@@ -0,0 +1,16 @@
+			"flags" : "BASIC_EXAMPLE=n IIO_EXAMPLE=y TARGET=max32690"
+		}
+	}
+}
```
new line at the eof missing

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,131 @@
+{
+	int ret, data, sign = 1;
+
+	/** Maximul value of 3V. */
```
looks like there is a typo in the comment

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,131 @@
+	if (ret)
+		goto exit;
+
+	/** Creating a triangle waveform. */
+	/** Initial value is 0. */
```
use the following comment format
``` 
/* line 1
 * line 2
 */
```

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,131 @@
+
+exit:
+	max2201x_remove(max2201x_desc);
+
```
how about no_os_gpio_remove for gpio0?

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,131 @@
+exit:
+	max2201x_remove(max2201x_desc);
+
+	return ret;
```
it would be nice to also print an error message before returning

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,131 @@
+		if (data <= 0)
+			sign = 1;
+		else if (data >= max_value)
+			sign = -1;
```
if / else if block needs the final else condition if / else if / else

rbolboac commented on projects/max2201x/src/platform/maxim/parameters.h:
```diff
@@ -0,0 +1,92 @@
+#elif (TARGET_NUM == 32665)
+#define UART_IRQ_ID		UART1_IRQn
+#define UART_DEVICE_ID		1
+#define UART_BAUDRATE		57600
```
can be taken out of the if macro seeing how it is common for both, furthermore you can move all specific defines in this if elif statement (including the spi and gpio stuff)

rbolboac commented on drivers/analog-io/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,919 @@
+	{
+		.name = "ldac_available",
+		.show = max2201x_iio_read_enable_available,
+		.shared = IIO_SHARED_BY_ALL
```
the shared field is not relevant for a debug attribute (only relevant for channel attributes, should be removed everywhere where applicable)

rbolboac commented on drivers/analog-io/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,919 @@
+	},
+	{
+		.name = "offset",
+		.store = max2201x_iio_write_offset,
```
any reason why .show is unavailable?

rbolboac commented on drivers/analog-io/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,919 @@
+
+	switch (priv) {
+	case MAX2201X_IIO_SLEW_RATE:
+		for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_enable_avail); i++)
+			if (!strcmp(buf, max2201x_enable_avail[i]))
+				break;
+
+		if (i == NO_OS_ARRAY_SIZE(max2201x_enable_avail))
+			return -EINVAL;
```
it looks like this behavior is repeated a number of times in this switch case, can you create a function to call each type with the according array and array size?

rbolboac commented on drivers/analog-io/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,919 @@
+static int max2201x_iio_read_step_size_available(void *dev, char *buf,
+		uint32_t len, const struct iio_ch_info *channel, intptr_t priv)
+{
+	uint32_t length = 0;
+	uint32_t i;
+
+	for (i = 0; i < NO_OS_ARRAY_SIZE(max2201x_step_size_avail); i++)
+		length += sprintf(buf + length, "%s ", max2201x_step_size_avail[i]);
```
this looks like it is being repeated a few times, maybe also this can be replaced with a function which takes as parameter the array, its size, the buffer and returns the length

rbolboac reviewed (commented):
some few more remarks from my side

RaduSabau1 commented:
V5

- Removed unrelated changes to commits.
- Added functions in iio driver to be used for reading available arrays and their lengths.
- Updated error label and solved memory leak in basic example.
- Fixed typos.
- Addressed other comments.

CiprianRegus commented on doc/sphinx/source/projects_doc.rst:
```diff
@@ -63,3 +63,10 @@ POWER MANAGEMENT
 
    projects/adp1050
+
+ANALOG-IO
```
What's the difference between "analog-io" and a DAC in this case?

RaduSabau1 commented:
V6

- Moved driver files and .rst documentation file to dac (more appropriate than creating a new folder).
- Changed path in the project makefiles.

RaduSabau1 commented:
V7

- Changed documentation path and type for both projec and driver.

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,138 @@
+
+remove_max2201x:
+	max2201x_remove(max2201x_desc);
+exit:
+	no_os_gpio_remove(gpio0);
```
this should be above remove_max2201x

rbolboac commented on projects/max2201x/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,138 @@
+
+	ret = max2201x_init(&max2201x_desc, &max2201x_ip);
+	if (ret)
+		return ret;
```
I would not return here directly without printing the error, please rework the error handling

rbolboac commented on drivers/dac/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,824 @@
+	return max2201x_reg_write(iio_desc->max2201x_desc, reg, writeval);
+}
+
+int max2201x_read_avail(char *array_avail, char *buf)
```
rename to max2201x_iio all defined APIs in this file

rbolboac reviewed (commented):
some remarks from my side, also please fix all build warnings

RaduSabau1 commented:
V8

- Reworked error handling in 'basic_example' file.
- Added '_iio' in 'read_avail' and 'read_avail_length' functions.

rbolboac commented:
you still have build warnings, please fix them
```
2024-03-25T10:53:24.2916931Z CC iio_max2201x.c
2024-03-25T10:53:24.3139448Z dac/max2201x/iio_max2201x.c: In function 'max2201x_iio_write_slew_rate':
2024-03-25T10:53:24.3144583Z dac/max2201x/iio_max2201x.c:504:12: warning: implicit declaration of function 'max2201x_iio_read_avail'; did you mean 'max2201x_iio_read_raw'? [-Wimplicit-function-declaration]
2024-03-25T10:53:24.3145907Z   504 |   enable = max2201x_iio_read_avail(max2201x_enable_avail, buf);
2024-03-25T10:53:24.3146629Z       |            ^~~~~~~~~~~~~~~~~~~~~~~
2024-03-25T10:53:24.3146877Z       |            max2201x_iio_read_raw
2024-03-25T10:53:24.3152005Z dac/max2201x/iio_max2201x.c:486:16: warning: unused variable 'i' [-Wunused-variable]
2024-03-25T10:53:24.3152533Z   486 |  uint32_t val, i;
2024-03-25T10:53:24.3152913Z       |                ^
2024-03-25T10:53:24.3162949Z dac/max2201x/iio_max2201x.c: In function 'max2201x_iio_read_step_size_available':
2024-03-25T10:53:24.3167447Z dac/max2201x/iio_max2201x.c:527:9: warning: implicit declaration of function 'max2201x_iio_read_avail_length'; did you mean 'max2201x_iio_read_offset'? [-Wimplicit-function-declaration]
2024-03-25T10:53:24.3168083Z   527 |  return max2201x_iio_read_avail_length(max2201x_step_size_avail, buf);
2024-03-25T10:53:24.3168694Z       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-03-25T10:53:24.3168917Z       |         max2201x_iio_read_offset
2024-03-25T10:53:24.3175695Z dac/max2201x/iio_max2201x.c: In function 'max2201x_iio_write_ao_config':
2024-03-25T10:53:24.3179561Z dac/max2201x/iio_max2201x.c:574:16: warning: unused variable 'i' [-Wunused-variable]
2024-03-25T10:53:24.3180091Z   574 |  uint32_t val, i;
2024-03-25T10:53:24.3180479Z       |                ^
2024-03-25T10:53:24.3188531Z CC max2201x.c
2024-03-25T10:53:24.3193821Z In file included from dac/max2201x/iio_max2201x.c:44:
2024-03-25T10:53:24.3194542Z dac/max2201x/iio_max2201x.c: In function 'max2201x_iio_read_avail':
2024-03-25T10:53:24.3200617Z ../include/no_os_util.h:54:13: warning: division 'sizeof (char *) / sizeof (char)' does not compute the number of array elements [-Wsizeof-pointer-div]
2024-03-25T10:53:24.3201206Z    54 |  (sizeof(x) / sizeof((x)[0]))
2024-03-25T10:53:24.3201697Z       |             ^
2024-03-25T10:53:24.3202232Z dac/max2201x/iio_max2201x.c:734:18: note: in expansion of macro 'NO_OS_ARRAY_SIZE'
2024-03-25T10:53:24.3202475Z   734 |  for (i = 0; i < NO_OS_ARRAY_SIZE(array_avail); i++)
2024-03-25T10:53:24.3202661Z       |                  ^~~~~~~~~~~~~~~~
2024-03-25T10:53:24.3207579Z dac/max2201x/iio_max2201x.c:730:35: note: first 'sizeof' operand was declared here
2024-03-25T10:53:24.3208061Z   730 | int max2201x_iio_read_avail(char *array_avail, char *buf)
2024-03-25T10:53:24.3208489Z       |                             ~~~~~~^~~~~~~~~~~
2024-03-25T10:53:24.3213877Z dac/max2201x/iio_max2201x.c:735:31: warning: passing argument 2 of 'strcmp' makes pointer from integer without a cast [-Wint-conversion]
2024-03-25T10:53:24.3214379Z   735 |   if (!strcmp(buf, array_avail[i]))
2024-03-25T10:53:24.3214793Z       |                    ~~~~~~~~~~~^~~
2024-03-25T10:53:24.3215194Z       |                               |
2024-03-25T10:53:24.3215394Z       |                               char
2024-03-25T10:53:24.3220657Z In file included from dac/max2201x/iio_max2201x.c:41:
2024-03-25T10:53:24.3221324Z /usr/include/newlib/string.h:36:28: note: expected 'const char *' but argument is of type 'char'
2024-03-25T10:53:24.3221784Z    36 | int  strcmp (const char *, const char *);
2024-03-25T10:53:24.3222113Z       |                            ^~~~~~~~~~~~
2024-03-25T10:53:24.3229764Z In file included from dac/max2201x/iio_max2201x.c:44:
2024-03-25T10:53:24.3230517Z ../include/no_os_util.h:54:13: warning: division 'sizeof (char *) / sizeof (char)' does not compute the number of array elements [-Wsizeof-pointer-div]
2024-03-25T10:53:24.3230927Z    54 |  (sizeof(x) / sizeof((x)[0]))
2024-03-25T10:53:24.3231430Z       |             ^
2024-03-25T10:53:24.3232168Z dac/max2201x/iio_max2201x.c:738:11: note: in expansion of macro 'NO_OS_ARRAY_SIZE'
2024-03-25T10:53:24.3232482Z   738 |  if (i == NO_OS_ARRAY_SIZE(array_avail))
2024-03-25T10:53:24.3232661Z       |           ^~~~~~~~~~~~~~~~
2024-03-25T10:53:24.3239881Z dac/max2201x/iio_max2201x.c:730:35: note: first 'sizeof' operand was declared here
2024-03-25T10:53:24.3240681Z   730 | int max2201x_iio_read_avail(char *array_avail, char *buf)
2024-03-25T10:53:24.3241021Z       |                             ~~~~~~^~~~~~~~~~~
2024-03-25T10:53:24.3244122Z In file included from dac/max2201x/iio_max2201x.c:44:
2024-03-25T10:53:24.3244637Z dac/max2201x/iio_max2201x.c: In function 'max2201x_iio_read_avail_length':
2024-03-25T10:53:24.3247432Z ../include/no_os_util.h:54:13: warning: division 'sizeof (char *) / sizeof (char)' does not compute the number of array elements [-Wsizeof-pointer-div]
2024-03-25T10:53:24.3247850Z    54 |  (sizeof(x) / sizeof((x)[0]))
2024-03-25T10:53:24.3248289Z       |             ^
2024-03-25T10:53:24.3248670Z dac/max2201x/iio_max2201x.c:749:18: note: in expansion of macro 'NO_OS_ARRAY_SIZE'
2024-03-25T10:53:24.3249000Z   749 |  for (i = 0; i < NO_OS_ARRAY_SIZE(array_avail); i++)
2024-03-25T10:53:24.3249268Z       |                  ^~~~~~~~~~~~~~~~
2024-03-25T10:53:24.3251889Z dac/max2201x/iio_max2201x.c:744:42: note: first 'sizeof' operand was declared here
2024-03-25T10:53:24.3252266Z   744 | int max2201x_iio_read_avail_length(char *array_avail, char *buf)
2024-03-25T10:53:24.3252558Z       |                                    ~~~~~~^~~~~~~~~~~
```

CiprianRegus commented on drivers/dac/max2201x/iio_max2201x.c:
```diff
@@ -0,0 +1,824 @@
+	int i;
+
+	for (i = 0; i < NO_OS_ARRAY_SIZE(array_avail); i++)
+		if (!strcmp(buf, array_avail[i]))
```
Is this working properly?

RaduSabau1 replied:
This is my bad here, indeed it shouldn't be working properly, my guess when I last tested it, it may have worked by coincidence and treated the result as a good one. Will test again with the hardware connected and make sure this function works properly in all the cases and the if condition doesn't check by accident.

RaduSabau1 commented:
V9

- Fixed bug in `max2201x_iio_read_avail` function.
- Reworked the way `_available` properties are read having them to use one common API and a specific `priv`.

RaduSabau1 commented:
V9

- Fixed bug in 'max2201x_iio_read_avail' function.
- Reworked the way '_available properties' are read having them to use one common API and a specific 'priv'.

RaduSabau1 commented:
V10

- Rebased with main.
- Included `dma` related files in the project's Makefiles used by the `no_os_irq` files according to last main branch update.
